### ISSUE TITLE
keepalive timer is ghosted and keeps running beyond connection lost, i think forever

### ISSUE THREAD

**Issue by tommyvn** [ronf/asyncssh#278]
`SSHConnection._set_keepalive_timer` is called multiple times in a connections lifetime.

The first time it's called `self._keepalive_timer` is set with the timer object.

Subsequent calls overwrite `self._keepalive_timer`, rendering previous instances of `self._keepalive_timer` un-cancellable.

https://github.com/ronf/asyncssh/blob/ef96070cf4eb801dc1a495caa1812fa039bb2854/asyncssh/connection.py#L510-L511 is the affected code.

I think it's as simple as making the check `if self._keepalive_interval and getattr(self, "_keepalive_timer", None) is None` but I don't at this moment have time to think properly about this so I'm recording the issue here before I forget.

**Comment by tommyvn** [comment]
I've monkey patched this into my server with close to zero thought and so far it's fixed my infinite keepalive issue:

```
class SSHServer(asyncssh.SSHServer):
    def connection_made(self, conn):
        def _set_keepalive_timer():
            if getattr(conn, "_keepalive_timer", None) is None:
                conn._set_keepalive_timer_orig()
        conn._set_keepalive_timer_orig = conn._set_keepalive_timer
        conn._set_keepalive_timer = _set_keepalive_timer
```

**Comment by ronf** [comment]
The code is designed right now to call `_reset_keepalive_timer()` whenever the existing timer might already be active. This function calls `_cancel_keepalive_timer()` before calling `_set_keepalive_timer()`, so any previous timer should be canceled before it is overwritten with a new value.

The only other calls to `_set_keepalive_timer()` are from within `_keepalive_timer_callback` when the timer actually goes off (so there shouldn't be a need to cancel the previous timer since it has already fired) and from `send_userauth_success` and `_process_userauth_success`, both of which should only be called before the keepalive timer has been started for the first time on a connection.

What behavior are you seeing that led you to file this issue?

As for your change, it looks to me like it would prevent the keepalive timer from being rescheduled from within `_keepalive_timer_callback` when it went off, effectively breaking this functionality. It would never get a chance to increment the `_keepalive_count` more than once, and so it would never reach the limit and kill the connection when it failed to get a keepalive response.

**Comment by tommyvn** [comment]
I turned asyncssh debug logging on.

I turn keepalives on by calling `conn.set_keepalive(interval=60)` in `asyncssh.SSHServer.connection_made`.

I ssh in and I see logs for keepalives for the duration of and beyond the termination of my ssh connection. The keepalive log messages for each connection continue until I terminate the server.

```
2020-05-05 12:44:54,238 - asyncssh.log:79 - DEBUG - [conn=0] Sending keepalive request
2020-05-05 12:44:54,243 - asyncssh.log:79 - DEBUG - [conn=0] Got keepalive response
2020-05-05 12:44:54,336 - asyncssh.log:79 - DEBUG - [conn=0] Sending keepalive request
2020-05-05 12:44:54,340 - asyncssh.log:79 - DEBUG - [conn=0] Got keepalive response
2020-05-05 12:44:54,894 - asyncssh.log:79 - INFO - [conn=0, chan=0] Sending exit status 0
2020-05-05 12:44:54,896 - asyncssh.log:79 - INFO - [conn=0, chan=0] Closing channel
2020-05-05 12:44:54,899 - asyncssh.log:79 - INFO - [conn=0, chan=0] Received channel close
2020-05-05 12:44:54,901 - asyncssh.log:79 - DEBUG - [conn=0] Received disconnect: disconnected by user (11)
2020-05-05 12:44:54,902 - asyncssh.log:79 - INFO - [conn=0, chan=0] Channel closed
2020-05-05 12:44:54,902 - asyncssh.log:79 - INFO - [conn=0] 0 connected clients
2020-05-05 12:44:56,244 - asyncssh.log:79 - DEBUG - [conn=0] Sending keepalive request
2020-05-05 12:44:56,244 - asyncssh.log:79 - DEBUG - [conn=0] Got keepalive response
2020-05-05 12:44:58,247 - asyncssh.log:79 - DEBUG - [conn=0] Sending keepalive request
2020-05-05 12:44:58,247 - asyncssh.log:79 - DEBUG - [conn=0] Got keepalive response
```

I discovered this because the amount of logs I received from my server after enabling keepalives were increasing linearly over time, this was the cause.

Perhaps calling conn.set_keepalive(interval=60) from within the SSHServer class is unexpected and I should be doing something else to fire it from within `send_userauth_success` or `_process_userauth_success` rather?

| it looks to me like it would prevent the keepalive timer from being rescheduled
I think what I'm proposing is a safety net rather than something that would interfere with rescheduling within _keepalive_timer_callback. It might be an oversimplification but it seems that if `_keepalive_timer` is already set it shouldn't be set again, or you risk rendering the underlying timer inaccessible but still running regardless of the timer checks.

This is certainly what's happening to me, I've hacked in enough logs to see two calls to `_set_keepalive_timer` and the subsequent ghosting, altho maybe because my path to setting keepalives is incorrect, I dunno, when I've got time I'll fire up pdb and follow the calls in a normal connection setup vs me calling `conn.set_keepalive` myself on connection to see properly.

**Comment by ronf** [comment]
Ah, yes - that could explain it. The keepalive isn't supposed to begin until after authentication completes. If you call `set_keepalive()` before authentication completes, it will enable the keepalive earlier than it should right now, and later when auth completes it will schedule it again, causing two keepalive timers to be running at once and cancellation won't work properly on them.

I should definitely add protective code to cover this case. I think it should just be a matter of the following:

```patch
diff --git a/asyncssh/connection.py b/asyncssh/connection.py
index 8a8ecda..3a61056 100644
--- a/asyncssh/connection.py
+++ b/asyncssh/connection.py
@@ -514,8 +514,9 @@ class SSHConnection(SSHPacketHandler, asyncio.Protocol):
     def _reset_keepalive_timer(self):
         """Reset the keepalive timer"""
 
-        self._cancel_keepalive_timer()
-        self._set_keepalive_timer()
+        if self._auth_complete:
+            self._cancel_keepalive_timer()
+            self._set_keepalive_timer()
 
     async def _make_keepalive_request(self):
         """Send keepalive request"""
@@ -764,8 +765,7 @@ class SSHConnection(SSHPacketHandler, asyncio.Protocol):
 
         self._inpbuf += data
 
-        if self._auth_complete:
-            self._reset_keepalive_timer()
+        self._reset_keepalive_timer()
 
         # pylint: disable=broad-except
```

I'm curious, though. Why do you set the keepalive timer in `connection_made()`, rather than setting it using the `keepalive_interval` argument when you are creating the listener for the SSHServer? If you do that, it'll automatically activate the keepalive at the right time (when auth completes).

If you need a timer to deal with dead connections before authentication completes, you're probably better off using the `login_timeout` for that. That will also prevent clients from connecting and keeping the TCP connection alive but not actually finishing authentication. This is enabled by default, with a value of 2 minutes.

**Comment by tommyvn** [comment]
> I'm curious, though. Why do you set the keepalive timer in connection_made()

TBH because I found the method on the conn object in the docs and then stopped looking. I'm not using it for anything as fancy as login timeouts, literally just wanted protocol level keepalives and that was the first way I found to enable them.

BTW I've switched to the `keepalive_interval` arg and it's working exactly as expected, thanks for that hint.
